proxy/src/metrics.rs

```diff
@@ -20,159 +29,203 @@ const DEFAULT_HTTP_REPORTING_TIMEOUT: Duration = Duration::from_secs(60);
-                if curr_val > prev_val {
-                    value = curr_val - prev_val;
-                    start_time = *prev_time;
-                } else {
-                    if curr_val < prev_val {
-                        error!("proxy_io_bytes_per_client metric value decreased from {} to {} for key {:?}",
-                        prev_val, curr_val, curr_key);
-                    }
-                    return None;
-                }
-            };
-
-            Some(Event {
-                kind: EventType::Incremental {
-                    start_time,
-                    stop_time: *curr_time,
-                },
-                metric: PROXY_IO_BYTES_PER_CLIENT,
-                idempotency_key: idempotency_key(hostname.to_owned()),
-                value,
-                extra: Ids {
-                    endpoint_id: curr_key.endpoint_id.clone(),
-                    branch_id: curr_key.branch_id.clone(),
-                },
-            })
-        })
-        .collect();
+    // only call counter_queue once.
+    // if we call this repeatedly, we might never exit the while loop below
+    // since we re-insert into this queue.
+    let mut counters = counter_queue.len();
 
-    if metrics_to_send.is_empty() {
-        trace!("no new metrics to send");
-        return Ok(());
-    }
+    info!(counters, "metrics sweep");
+
+    while counters > 0 {
+        let mut i = 0;
+        while i < events.len() && i < counters {
+            let Some(counter) = counter_queue.pop_front() else { break };
+
+            // update metrics eagerly. If there's a failure, there is a chance that this metric won't get recorded.
+            // We don't want to send metrics twice as it might cause double billing which is worse than not billing enough.
+            // see the relevant discussion: https://github.com/neondatabase/neon/pull/4563#discussion_r1246710956
+            let tx = counter.tx.swap(0, Ordering::Relaxed);
+            let _rx = counter.rx.swap(0, Ordering::Relaxed);
+
+            // only record the metric event if it's not 0
+            if tx > 0 {
+                let now = Utc::now();
+                let last_seen = Utc.timestamp_nanos(
+                    counter
+                        .last_seen
+                        .swap(now.timestamp_nanos(), Ordering::Relaxed),
+                );
+
+                // try to use minimal allocations
+                let event = &mut events[i];
+                event.extra.clone_from(&counter.ids);
+                idempotency_key_into(&hostname, &mut event.idempotency_key);
+
+                // Only collect metric for outbound traffic
+                event.value = tx;
+                event.metric = PROXY_IO_BYTES_PER_CLIENT;
+                event.kind = EventType::Incremental {
+                    start_time: last_seen,
+                    stop_time: now,
+                };
+
+                i += 1;
+            } else {
+                counters -= 1;
```

Comment:
Does this mean that we will drop the counters for a particular sometime after we've last uploaded the values (because tx does not change, we end up here)?